### PR TITLE
🏗 Parallelize visual diff tests in multiple browser tabs

### PR DESCRIPTION
**Pull request by danielrozenberg**
This should speed up processing time without increasing computation by much, since most of the time is wasted on waiting for navigation to finish.

Unfortunately when I run these locally I'm getting flakes, e.g. see these two builds:
* https://percy.io/ampproject/danielrozenberg/builds/892614
* https://percy.io/ampproject/danielrozenberg/builds/892649

I'm not sure why these flakes appear now but didn't appear before. e.g., the video loading progress bar in `amp-video`. Any guesses as to why those happen now but didn't happen before?

I'll let this run on Travis as well, maybe Travis will be better :)

### CODE REVIEW

**Comment by rsimha** [comment]
I'll review this tomorrow, but a question in the meantime: Does this take into account the needs of different tests, where each one has its own AMP config, experiments, etc.? Aren't we using one server to serve the runtime to all these pages?

**Comment by danielrozenberg** [comment]
All pages using the same config will execute in the same batch, until that config is "done", then the next config starts.

You are correct that I didn't take the experiments into account correctly, I'll have to think how I can solve this...

**Comment by rsimha** [comment]
IIRC, experiments are entirely governed by cookies for the browser instance, so you could potentially get away with using one server and multiple browser instances, each using its own cookie jar.

**Comment by danielrozenberg** [comment]
After discussing this offline with @rsimha: I can't set experiements using the # fragment part of the URL because that's only relevant to `<amp-experiment>` components, not for global experiments (confusing :disappointed:)

However, there is currently only one visual diff test that uses the experiments field, and it seems to be unnecessary since that feature has launched and is no longer behind an experiment flag.

I'm removing the specialized support for experiments from `visual-diff.js`. As an alternative, one can set experiments using a `<meta>` tag. See: https://www.ampproject.org/docs/reference/experimental#enable-an-experiment-for-a-particular-document

**Comment by rsimha** [comment]
@danielrozenberg I haven't looked at this in a while. Has it gotten to a working state, and is it ready for review?

**Comment by danielrozenberg** [comment]
This slipped through the cracks for me. It's mostly ready for review, but I still need to figure out why it works locally but fails on Travis about half the times

File: build-system/tasks/visual-diff.js [hunk 1]
```diff
@@ -41,6 +41,8 @@ const PORT = 8000;
 const BASE_URL = `http://${HOST}:${PORT}`;
 const WEBSERVER_TIMEOUT_RETRIES = 10;
 const NAVIGATE_TIMEOUT_MS = 3000;
+const MAX_PARALLEL_PAGES = 10;
```
Comment (rsimha) [inline build-system/tasks/visual-diff.js]: Based on the comment below, how about calling this `AVAILABLE_TABS`?

File: build-system/tasks/visual-diff.js [hunk 2]
```diff
@@ -422,51 +417,68 @@ async function snapshotWebpages(percy, page, webpages) {
...
+                .replace(/__HEIGHT__/g, viewport.height));
+            await page.setViewport({
+              width: VIEWPORT_WIDTH,
+              height: VIEWPORT_HEIGHT,
+            });
+          }
+
+          await percy.snapshot(name, page, snapshotOptions);
+          await page.close();
+          log('travis', colors.cyan('●'));
```
Comment (rsimha) [inline build-system/tasks/visual-diff.js]: nit: I've noticed that when an error is encountered, it gets printed inline with these dots. Easy fix: insert a newline before any errors printed during tests.

File: build-system/tasks/visual-diff.js [hunk 3]
```diff
@@ -706,15 +692,16 @@ async function visualDiff() {
...
     log('fatal', `Failed to start a web server: ${reason}`);
   });
 
   if (argv.skip) {
+    const browser = await launchBrowser();
+    const page = await newPage(browser);
     await createEmptyBuild(page);
-    await shutdown(webServerProcess);
+    process.exit(0);
     return;
```
Comment (rsimha) [inline build-system/tasks/visual-diff.js]: `return` after `process.exit(0)` is moot.

File: build-system/tasks/visual-diff.js [hunk 4]
```diff
@@ -41,6 +41,8 @@ const PORT = 8000;
 const BASE_URL = `http://${HOST}:${PORT}`;
 const WEBSERVER_TIMEOUT_RETRIES = 10;
 const NAVIGATE_TIMEOUT_MS = 3000;
+const MAX_PARALLEL_PAGES = 10;
+const WAIT_FOR_PRODUCER_OR_CONSUMER_MS = 1000;
```
[the remark this answers is not in the document]
Comment (rsimha) [reply build-system/tasks/visual-diff.js]: SGTM.

File: build-system/tasks/visual-diff.js (hunk 3 above)
Comment (rsimha) [reply build-system/tasks/visual-diff.js]: I'm surprised it complained, but sure, best to avoid an alert.

**Comment by danielrozenberg** [comment]
I ran the Travis build for this PR ~20 times (after the last significant changes on here) to ensure that this doesn't introduce flakes. Looks good so far. Merging!

**Pull request merged**